File: app/src/main/java/kztproject/jp/splacounter/AuthRepository.java
package kztproject.jp.splacounter;

import javax.inject.Inject;

import io.reactivex.Completable;
import kztproject.jp.splacounter.api.MyServiceClient;
import kztproject.jp.splacounter.model.User;
import kztproject.jp.splacounter.model.UserResponse;
import kztproject.jp.splacounter.preference.AppPrefs;
import kztproject.jp.splacounter.preference.AppPrefsProvider;

public class AuthRepository {

    private final MyServiceClient client;
    private final AppPrefsProvider prefs;

    @Inject
    public AuthRepository(MyServiceClient client, AppPrefsProvider prefs) {
        this.client = client;
        this.prefs = prefs;
    }

    public Completable login(String inputString) {
        return client.getUser(inputString)
                .flatMapCompletable(this::save);
    }

    private Completable save(UserResponse response) {
        User user = response.user;
        AppPrefs schema = prefs.get();
        schema.putUserId(user.id);
        schema.putUserName(user.fullName);
        return Completable.complete();
    }
}

File: app/src/main/java/kztproject/jp/splacounter/auth/repository/AuthRepository.kt
package kztproject.jp.splacounter.auth.repository

import io.reactivex.Completable
import kztproject.jp.splacounter.api.RewardListClient
import kztproject.jp.splacounter.auth.api.TodoistService
import kztproject.jp.splacounter.model.RewardUser
import kztproject.jp.splacounter.model.UserResponse
import kztproject.jp.splacounter.preference.PrefsWrapper
import javax.inject.Inject

class AuthRepository @Inject constructor(private val todoistService: TodoistService,
                                         private val rewardListClient: RewardListClient) {

    fun login(inputString: String): Completable {
        return todoistService.getUser(inputString, "*", "[\"user\"]")
                .flatMap { response: UserResponse -> rewardListClient.findUser(response.user.id.toLong()) }
                .flatMapCompletable { save(it) }
    }

    fun signUp(todoistToken: String): Completable {
        return todoistService.getUser(todoistToken, "*", "[\"user\"]")
                .flatMap { response -> rewardListClient.createUser(response.user.id.toLong()) }
                .flatMapCompletable { save(it) }
    }

    private fun save(user: RewardUser): Completable {
        PrefsWrapper.userId = user.id
        return Completable.complete()
    }
}
